core/primitives/src/errors.rs

```diff
@@ -27,73 +29,144 @@ impl std::fmt::Display for StorageError {
 impl std::error::Error for StorageError {}
 
 /// External
-#[derive(BorshSerialize, BorshDeserialize, Debug, Clone, PartialEq, Eq)]
+#[derive(BorshSerialize, BorshDeserialize, Debug, Clone, PartialEq, Eq, Deserialize, Serialize)]
 pub enum InvalidTxError {
-    InvalidSigner(AccountId),
-    SignerDoesNotExist(AccountId),
     InvalidAccessKey(InvalidAccessKeyError),
-    InvalidNonce(Nonce, Nonce),
-    InvalidReceiver(AccountId),
+    InvalidSigner {
+        signer_id: AccountId,
+    },
+    SignerDoesNotExist {
+        signer_id: AccountId,
+    },
+    InvalidNonce {
+        tx_nonce: Nonce,
+        ak_nonce: Nonce,
+    },
+    InvalidReceiver {
+        receiver_id: AccountId,
+    },
     InvalidSignature,
-    NotEnoughBalance(AccountId, Balance, Balance),
-    RentUnpaid(AccountId, Balance),
+    NotEnoughBalance {
+        signer_id: AccountId,
+        #[serde(with = "u128_dec_format")]
+        balance: Balance,
+        #[serde(with = "u128_dec_format")]
+        cost: Balance,
+    },
+    RentUnpaid {
+        signer_id: AccountId,
+        #[serde(with = "u128_dec_format")]
+        amount: Balance,
+    },
     CostOverflow,
     InvalidChain,
     Expired,
 }
 
-#[derive(BorshSerialize, BorshDeserialize, Debug, Clone, PartialEq, Eq)]
+#[derive(BorshSerialize, BorshDeserialize, Debug, Clone, PartialEq, Eq, Deserialize, Serialize)]
 pub enum InvalidAccessKeyError {
-    AccessKeyNotFound(AccountId, PublicKey),
-    ReceiverMismatch(AccountId, AccountId),
-    MethodNameMismatch(String),
+    AccessKeyNotFound {
+        account_id: AccountId,
+        public_key: PublicKey,
+    },
+    ReceiverMismatch {
+        tx_receiver: AccountId,
+        ak_receiver: AccountId,
+    },
+    MethodNameMismatch {
+        method_name: String,
+    },
     ActionError,
-    NotEnoughAllowance(AccountId, PublicKey, Balance, Balance),
+    NotEnoughAllowance {
+        account_id: AccountId,
+        public_key: PublicKey,
+        #[serde(with = "u128_dec_format")]
+        allowance: Balance,
+        #[serde(with = "u128_dec_format")]
+        cost: Balance,
+    },
 }
 
-#[derive(BorshSerialize, BorshDeserialize, Debug, Clone, PartialEq, Eq)]
+#[derive(BorshSerialize, BorshDeserialize, Debug, Clone, PartialEq, Eq, Deserialize, Serialize)]
 pub enum ActionError {
-    AccountAlreadyExists(AccountId),
-    AccountDoesNotExist(String, AccountId),
-    CreateAccountNotAllowed(AccountId, AccountId),
-    ActorNoPermission(AccountId, AccountId, String),
-    DeleteKeyDoesNotExist(AccountId),
-    AddKeyAlreadyExists(PublicKey),
-    DeleteAccountStaking(AccountId),
-    DeleteAccountHasRent(AccountId, Balance),
-    RentUnpaid(AccountId, Balance),
-    TriesToUnstake(AccountId),
-    TriesToStake(AccountId, Balance, Balance, Balance),
-    FunctionCallError(String), // TODO type
+    AccountAlreadyExists {
+        account_id: AccountId,
+    },
+    AccountDoesNotExist {
+        action: String,
```

Comment:
Instead of returning action type converting to string we should return the action index.

Action type does not allow us to recover which action specifically caused the error because there might be multiple actions of the same type in the transaction.
Also, we cannot attach the entire action here, because some actions are too large (e.g. contract deployment).

Reply:
Yeah, you right. As you can see I left the general error structure unchanged (except that I switched from tuple variants to struct variants). So, my plan is to change the structure in the next iteration. I think it makes sense to change ActionError to struct and have an "index" and "kind" fields for the actual action (set the current enum to).

Reply:
adressed in https://github.com/nearprotocol/nearcore/pull/1854/commits/30622910a44d9eadfa22b757b7a54ac0936d2559

Reply:
> So, my plan is to change the structure in the next iteration.

Splitting it into several PRs is more preferable.

Reply:
* commits